File: tests/synthetic.py
"""Synthetic rasters for tests and examples; no network needed."""

import os

import numpy as np
from osgeo import gdal, osr

gdal.UseExceptions()


def make_tile(path, nx, ny, block, origin, res, seed=0, nodata=None,
              value_offset=0.0, nodata_frac=0.02):
    drv = gdal.GetDriverByName("GTiff")
    ds = drv.Create(path, nx, ny, 1, gdal.GDT_Float32,
                    options=["TILED=YES", f"BLOCKXSIZE={block[0]}",
                             f"BLOCKYSIZE={block[1]}", "COMPRESS=DEFLATE"])
    ds.SetGeoTransform((origin[0], res, 0, origin[1], 0, -res))
    srs = osr.SpatialReference()
    srs.ImportFromEPSG(4326)
    ds.SetProjection(srs.ExportToWkt())
    rng = np.random.default_rng(seed)
    r = np.arange(ny, dtype=np.float32)[:, None]
    c = np.arange(nx, dtype=np.float32)[None, :]
    data = (value_offset + r * 1000 + c).astype(np.float32)
    band = ds.GetRasterBand(1)
    if nodata is not None:
        band.SetNoDataValue(nodata)
        data[rng.random((ny, nx)) < nodata_frac] = nodata
    band.WriteArray(data)
    ds = None
    return path


def make_mosaic(base, layout=(3, 2), size=(300, 200), blocks=((64, 64),),
                overlap=(0, 0), drop=(), nodata=None, res=0.01,
                origin=(100.0, -30.0), vrt_nodata=None):
    """Tiles laid out on one grid plus a VRT over them. Returns the VRT path.

    Tile k (row-major in the layout) holds values k * 1e6 + row * 1000 + col.
    """
    os.makedirs(base, exist_ok=True)
    nxl, nyl = layout
    sx, sy = size
    paths = []
    k = 0
    for j in range(nyl):
        for i in range(nxl):
            if k not in drop:
                ox = origin[0] + i * (sx - overlap[0]) * res
                oy = origin[1] - j * (sy - overlap[1]) * res
                p = os.path.join(base, f"tile_{j}_{i}.tif")
                make_tile(p, sx, sy, blocks[k % len(blocks)], (ox, oy), res,
                          seed=k, nodata=nodata, value_offset=k * 1e6)
                paths.append(p)
            k += 1
    vrt = os.path.join(base, "mosaic.vrt")
    if vrt_nodata is None:
        vrt_nodata = nodata if nodata is not None else "None"
    opts = gdal.BuildVRTOptions(VRTNodata=vrt_nodata,
                                srcNodata=nodata if nodata is not None else "None")
    gdal.BuildVRT(vrt, paths, options=opts).FlushCache()
    return vrt

File: tests/synthetic.R
## Synthetic rasters for R tests and examples; no network needed.
## Tile k holds values k * 1e6 + row * 1000 + col (0-based row/col in the value
## only, so the numbers match the Python fixtures in synthetic.py).

td <- tempfile("pixtract"); dir.create(td)

make_tile <- function(path, nx, ny, block, origin, res, k, nodata = NA) {
  ds <- create("GTiff", path, nx, ny, 1L, "Float32", return_obj = TRUE,
               options = c("TILED=YES", paste0("BLOCKXSIZE=", block[1L]),
                           paste0("BLOCKYSIZE=", block[2L]), "COMPRESS=DEFLATE"))
  ds$setGeoTransform(c(origin[1L], res, 0, origin[2L], 0, -res))
  ds$setProjection(srs_to_wkt("EPSG:4326"))
  ## raster order: row-major from the top-left
  v <- k * 1e6 + rep((seq_len(ny) - 1) * 1000, each = nx) + rep(seq_len(nx) - 1, ny)
  if (!is.na(nodata)) {
    set.seed(k)
    v[runif(length(v)) < 0.02] <- nodata
    ds$setNoDataValue(1L, nodata)
  }
  ds$write(1L, 0L, 0L, nx, ny, v)
  ds$close()
  path
}

make_mosaic <- function(name, layout = c(3, 2), size = c(300, 200), blocks = list(c(64, 64)),
                        overlap = c(0, 0), drop = integer(), nodata = NA, vrt_nodata = nodata) {
  base <- file.path(td, name); dir.create(base)
  paths <- character(); k <- 0
  for (j in seq_len(layout[2L]) - 1) for (i in seq_len(layout[1L]) - 1) {
    if (!k %in% drop) {
      o <- c(100 + i * (size[1L] - overlap[1L]) * 0.01, -30 - j * (size[2L] - overlap[2L]) * 0.01)
      p <- file.path(base, sprintf("tile_%d_%d.tif", j, i))
      paths <- c(paths, make_tile(p, size[1L], size[2L], blocks[[k %% length(blocks) + 1]],
                                  o, 0.01, k, nodata))
    }
    k <- k + 1
  }
  vrt <- file.path(base, "mosaic.vrt")
  args <- c("-vrtnodata", if (is.na(vrt_nodata)) "None" else vrt_nodata,
            "-srcnodata", if (is.na(nodata)) "None" else nodata)
  buildVRT(vrt, paths, cl_arg = args, quiet = TRUE)
  vrt
}
